Code review of RFC - Simulation Interfaces (#2)
RobotecAI/sig-simulation: rfcs/RFC Simulation Interfaces.md

Reviewer: According to convention in the rfcs directory, I would go for the name:
"RFC Feature - Simulation Interfaces Integration.md"

### Summary:
<!-- Single paragraph explanation of the feature -->

> This RFC is a follow-up to an effort at [ros-simulation/simulation_interfaces](https://github.com/ros-simulation/simulation_interfaces/pull/1) to standardize simulation interfaces existing robotics simulators in [ROS 2](https://docs.ros.org/en/jazzy/index.html). Current implementation of similar interfaces is limited to `ROS2SpawnerComponent`, that utilizes *Gazebo* messages for ROS 2 communication. The messages were marked deprecated in the latest ROS 2 release and should be superseded. Proposed approach includes three new *Components* that will fulfill the design presented in [ros-simulation/simulation_interfaces](https://github.com/ros-simulation/simulation_interfaces/pull/1), and that are planned to be developed alongside updates to simulation interfaces. Backward compatibility will be ensured for `ROS2SpawnerComponent` and `ROS2ContactSensor` through CMake configuration with the deprecation information.

Reviewer: A minor rephrase to consider.
```suggestion
This RFC is a follow-up to an effort at [ros-simulation/simulation_interfaces](https://github.com/ros-simulation/simulation_interfaces/pull/1) to standardize [ROS 2](https://docs.ros.org/en/jazzy/index.html) simulation interfaces for robotics simulators. Current implementation of similar interfaces is limited to `ROS2SpawnerComponent`, that utilizes *Gazebo* messages for ROS 2 communication. The messages were marked deprecated in the latest ROS 2 release and should be superseded. Proposed approach includes three new *Components* that will fulfill the design presented in [ros-simulation/simulation_interfaces](https://github.com/ros-simulation/simulation_interfaces/pull/1), and that are planned to be developed alongside updates to simulation interfaces. Backward compatibility will be ensured for `ROS2SpawnerComponent` and `ROS2ContactSensor` through CMake configuration with the deprecation information.
```
+
I would also reconsider mentioning "three new Components", because it is highly possible there will be more of them. Maybe we can simply mention that there will be "three core components"?

The feature is an API for handling number of ROS 2 [services](https://docs.ros.org/en/jazzy/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Services/Understanding-ROS2-Services.html) and [actions](https://docs.ros.org/en/jazzy/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Actions/Understanding-ROS2-Actions.html).

The following terminology that was created in [RFC-410](https://github.com/ros-infrastructure/rep/issues/410) in [ros-infrastructure/rep](https://github.com/ros-infrastructure/rep):

| Term      | Description                                                                      |
| --------- | -------------------------------------------------------------------------------- |
| Spawnable | Robot or other object that can be spawned in simulation runtime.                 |
| Entity    | Spawned spawnable, it has a unique name.                                         |

> | Bound     | A region that is defined by an axis-aligned box shape, convex hull, or a sphere. |

Reviewer: ```suggestion
| Bounds     | A volume that is defined by an axis-aligned box shape, convex hull, or a sphere. |
```
In GetNamedPoseBounds we mention that
> Note: Convex hull bound shape will not be supported in this implementation. This subject will be cover by another RFC when necessary.

This is not present under GetEntityBounds. I write this here to clarify, if we would like to support convex hull bounds?

| NamedPose | SE3 (translation and rotation) transform with a unique name                      |
| Tag       | A string that allows filtering entities and named poses                          |

The implementation will be split into three (or more) system components:
- `ROS 2 Entities manager`: responsible for the lifetime of spawned objects, it will cache initial positions.
- `ROS 2 Named poses manager`: responsible for aggregating information in the Named Pose Game Component.
- `ROS 2 Simulator manager`: responsible for modifying the global state of the simulation (e.g., pausing, reloading).

We will decouple the implementation of those features from their ROS 2 interfaces. Every manager will expose public methods that:
- will be callable from C++,
- will be handled through dedicated ROS 2 interface and exposed as service.

> The purpose of that approach is to enable testability without the need for a ROS framework and ensure the whole system can be used with any middleware in the future. 

Reviewer: ```suggestion

The purpose of that approach is to enable testability without the need for a ROS framework and ensure the whole system can be used with any middleware in the future. 
```
Otherwise, this sentence is merge with the line above.

The simulator needs to advertise supported features via the ROS 2 service [GetSimulatorFeatures.srv](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/srv/GetSimulatorFeatures.srv)
That service in its response provides the caller with a list of features [SimulatorFeatures.msg](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/msg/SimulatorFeatures.msg)

The following features are the subject of this RFC:

>  - SPAWNING

Reviewer: https://github.com/adamdbrw/simulation_interfaces/pull/1 mentions also ENTITY_TAGS feature which seems to be not listed anywhere here. This should rather be added to this list, because we simply provide implementation that works with tags.

 - SIMULATION_PAUSE

> The features *SIMULATION_RESET_STATE*, *STEP_SIMULATION_SINGLE*, *STEP_SIMULATION_MULTIPLE*, *STEP_SIMULATION_ACTION* will be introduced in the next RFC since they require multiple changes in the *PhysX Gem* and *AzPhysics API*. Action [SimulateSteps.action](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/action/SimulateSteps.action) will not be supported.

Reviewer: SIMULATION_RESET_STATE is mentioned here, but is also listed above as a subject to this RFC. As far as I remember, we rather considered implementing this in current stage, but we may also discuss this further.

A simulation expert who wants to create a robot (or other simulation-ready asset) needs:
- design a new prefab using standard O3DE workflow, using all available components,
- save the prefab with a name e.g., `Robots/FooRobot.prefab`
- create and fill the `SimulationInfo` sidecar file as `Robots/FooRobot.SimulationInfo` using a text editor or the O3DE tooling
- export simulation with the export script and [bundle assets](https://docs.o3de.org/docs/user-guide/packaging/asset-bundler/)

A simulation interfaces user who calls the `GetSpawnables` service against GameLauncher will call the asset catalog to find all product assets of the type `.simulationinfo`. The `ROS 2 Entity manager` will aggregate found assets and will prepare a response that will contain:
   * `uri` as `spawnable://@cache@/robot/foorobot.spawnable`
   * the description copied from `Robots/FooRobot.SimulationInfo`,
   * the bound information copied from `Robots/FooRobot.SimulationInfo`,
   * tag list copied from `Robots/FooRobot.SimulationInfo`,
   * category copied from `Robots/FooRobot.SimulationInfo`.

> Prepared response will be returned to the ROS 2 user.

Reviewer: ```suggestion

Prepared response will be returned to the ROS 2 user.
```
Otherwise, merged with the line above.
+
Do we want to clarify, what will happen if some of this data will not be present in SimulationInfo?

- The [ROS2FrameComponent](https://www.docs.o3de.org/docs/user-guide/components/reference/ros2/core/ros2-frame/) will reflect the namespace the value of the `entity_namespace` field.
- A component called `SimulationInfoComponent` will be created and attached to the root entity to cache the information from `simulationinfo` and to make it easily accessible in the future.
- [TagComponent](https://www.docs.o3de.org/docs/user-guide/components/reference/gameplay/tag/) from LmbrCentral Gem will be created at root entity (if not present) and updated with a list of tags from the `simulationinfo` sidecar.

`SimulationInfoComponent` will be handling a `SimulationInfoComponentRequestBus`.
`SimulationInfoComponentRequestBus` will have a unique string key that will come from the name of spawned entity.
This approach will result in a centralized aggregation of entities' information, and will provide an API to find all spawned entities and get their states.

## GetEntities service

This service provides access to a list of all spawned entities. \
Service definition: [GetEntities](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/srv/GetEntities.srv )

> We will serve this by calling `SimulationInfoComponentRequestBus` with `AZ::EBusAggregateResults` to obtain a list of `AZ::EntityId`, that can be converted to a list of entity names.

Reviewer: Or maybe we will have some lookup in Entities manager? Because we may need these ids for different purposes.

Reviewer 2: I like to keep my component stateless, so please keep it as is.

Individual entity state definition: [EntityState](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/msg/EntityState.msg)

The service handling will be similar to [GetEntities](#GetEntities). Additional calls will be required to fill the state information:
 - [RigidBodyRequestBus](https://github.com/o3de/o3de/blob/79e1df3990c5554c454d68798a0f3ef94c8ae317/Code/Framework/AzFramework/AzFramework/Physics/RigidBodyBus.h) will be used to find the current twist,
 - [TransformComponentRequests](https://github.com/o3de/o3de/blob/42d375dd99b972400f9f26bfec7e3444088f3398/Code/Framework/AzCore/AzCore/Component/TransformBus.h) will be used to find the current pose,
 - [ROS2FrameBus.h](https://github.com/o3de/o3de-extras/blob/development/Gems/ROS2/Code/Include/ROS2/Frame/ROS2FrameBus.h) will be used to find the frame ID.

**Note:** Acceleration will not be filled, and will not be supported.

## SetEntityState service

This service allows modifying the state of the chosen entity. \
Service definition: [SetEntityState.srv](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/srv/SetEntityState.srv)

> The `AZ::EntityId` will be discovered by calling `SimulationInfoComponentRequestBus` with the name provided in the request. Next, the corresponding API will be called to adjust the state of the entity:

Reviewer: Similarly to my comment for GetEntities - this may also be another lookup. However, I am not sure if we want to go into such details at this point?

Reviewer 2: I want to avoid managing a state in Entities manager.

- Potential 3rd physics engine.

Some of those methods respect calls to `TranformComponentRequests`, but some of them need to call directly methods in the PhysX gem.
| Method                                  | TransformComponentRequests API | RigidBodyRequestBus API | Custom API |
| --------------------------------------- | ------------------------------ | ----------------------- | ---------- |
| Simulated PhysX Rigid Body              |                                | supports                |            |
| Kinematic PhysX Rigid Body              | supports                       |                         |            |
| TransformComponent  due to parent-child | supports                       |                         |            |
| TransformComponent (custom calls)       |                                |                         |            |

> | PhysX character component               |                                |                         | ?          |

Reviewer 2: ```suggestion
| PhysX character component               |                                |                         | TBD         |
```

> | PhysX character component               |                                |                         | ?          |

Reviewer 2: ```suggestion
| PhysX character component               |                                |                         | TBD        |
```

> | PhysX ragdoll                           |                                |                         | ?          |

Reviewer: These rows may need further explanation. I guess "?" means "has to be researched"?
+
I assume that we have to choose some solution for this RFC. Can we support only some subset of methods presented in this table? Lets discuss this.

Reviewer 2: I will left this discussion to public.

Author: Keeping it unchanged.

The velocity (twist) can be applied using `RigidBodyRequestBus::SetLinearVelocity` and `RigidBodyRequestBus::SetAngularVelocity`.
It will be effective against Simulated PhysX Rigid Body and Kinematic PhysX Rigid Body and should give a warning if it called against unsupported type.

The acceleration can be applied using `RigidBodyRequestBus::ApplyLinearImpulse` w.r.t object's mass.
It will be effective against Simulated PhysX Rigid Body and should give a warning if it is called against 
unsupported type.

**Note:** Currently ROS2 Gem is heavily dependent on PhysX5 (due to SDF importer). The decoupling of those needs to be done in the future.

## DeleteEntity service

This service allows despawning the previously spawned entities. \
Service definition: [DeleteEntity.srv](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/srv/DeleteEntity.srv)

The `AZ::EntityId` will be discovered by calling `SimulationInfoComponentRequestBus` with the name provided in the request. Next, the corresponding API will be called to despawn and remove the corresponding spawn ticket.

> **Note:** This mechanism will allow to delete the entities that are part of the level prefab (e.g., prefab instantiated in Editor).

Reviewer: Just to be sure if "now allow to delete" in previous version was not intended to be "not allow to delete". At some point we have discussed only entities spawned during simulation being visible through simulation interfaces. However, with using correctly prepared prefabs (SimulationInfo) it could be possible to detect corresponding entities to.

@michalpelka @jhanca-robotecai What is your opinion on this?

Author: It was clarified by @michalpelka that the system will *not* allow any changes in the level itself. I applied this in the text.

> **Note:** This mechanism will allow to delete the entities that are part of the level prefab (e.g., prefab instantiated in Editor).

Reviewer 2: ```suggestion
**Note:** This mechanism will **not** allow to delete the entities that are part of the level prefab (e.g., prefab instantiated in Editor).
```

## GetEntityBounds service

This service allows to get the bounds of the previously spawned entities. \
Service definition: [GetEntityBounds.srv](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/srv/GetEntityBounds.srv)

> The bound for the particular entity will be obtained by calling `SimulationInfoComponentRequestBus` with the name of the entity provided in the request. The information about entity's bounds will be stored within the `.simulationinfo` product asset, and it will be generated based on the source asset data provided by the simulation engineer.

Reviewer: ```suggestion
The bounds for the particular entity will be obtained by calling `SimulationInfoComponentRequestBus` with the name of the entity provided in the request. The information about entity's bounds will be stored within the `.simulationinfo` product asset, and it will be generated based on the source asset data provided by the simulation engineer.
```

## GetNamedPoses service

This service allows to get the list of the predefined poses which are convenient to for spawning, navigation goals, etc. \
Service definition: [GetNamedPoses.srv](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/srv/GetNamedPoses.srv) \
Individual named pose definition: [NamedPose.msg](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/msg/NamedPose.msg)

This information will be obtained by calling a respective bus of `ROS 2 Named poses manager`. Additionally, the aggregation will be filtered based on the parameters of the call. The poses will be predefined by a simulation expert using O3DE Editor. In particular, each entity with `NamedPoseComponent` from ROS 2 Gem will be registered in the poses manager. The `NamedPoseComponent` will contain two configurable fields called `description` and `tags`. During creation of game component (`CreateGameEntity`) the `TagComponent` from `LmbrCentral` Gem will be created and fed with the list of tags. 

**Note:** `ROS 2 Named poses manager` will ensure the names are unique (O3DE does not check for the name's uniqueness).

## GetNamedPoseBounds service

> This service allows to get the boundaries defined in the predefined pose object. \

Reviewer: This service description is the only place, where you refer to 'boundaries' instead of 'bounds', multiple times. May be worth unifying, although it is not very confusing.

> This service allows to get the boundaries defined in the predefined pose object. \

Reviewer 2: ```suggestion
This service allows to get the bounds defined in the predefined pose object. \
```

Service definition [GetNamedPoseBounds.srv](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/srv/GetNamedPoseBounds.srv) 

The information about the boundaries will be obtained by calling a respective bus of `ROS 2 Named poses manager`. Additionally, the aggregation will be filtered based on the parameters of the call. The boundaries of the pose will be predefined by a simulation expert using O3DE Editor by adding `TransformService` and dependent services `BoxShapeService` and `SphereShapeService` alongside with the `NamedPoseComponent`.

> **Note:** Convex hull bound shape will not be supported in this implementation. This subject will be cover by another RFC when necessary.

Reviewer: ```suggestion
**Note:** Convex hull bounds shape will not be supported in this implementation. This subject will be cover by another RFC when necessary.
```

## Reset Simulation service

This service allows to reset the simulation via ROS 2 interface. \
Service definition: [ResetSimulation.srv](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/srv/ResetSimulation.srv)

This service will be handled by `ROS 2 Simulator manager`. It will use multiple APIs to give results.

| Scope         | Planned API and usage                                                          |
| ------------- | ------------------------------------------------------------------------------ |
| SCOPE_ALL     | `ConsoleRequestBus` and `LoadLevel` command.                                   |
| SCOPE_SPAWNED | Internal API to destroy all spawn tickets using `ROS 2 Entities manager`.      |
| SCOPE_STATE   | Move all spawned entities to initial poses cached in `ROS 2 Entities manager`. |
| SCOPE_TIME    | New call using `ROS2Bus`                                                       |

> ## SetSimulationState service

Reviewer: It may be worth noting here or under GetSimulationState, what will be the initial state? According to simulation interfaces RFC simulation starts as PAUSED.

Moreover, I would at least consider, what is before this? I guess the ROS 2 simulation interfaces become available only when simulation is initialized, and the state is first set to PAUSED.

This service allows to set the state of the simulation (*STOPPED*, *PAUSED*, *PLAYING*, *QUITTING*). \
Service definition: [SimulationState](https://github.com/adamdbrw/simulation_interfaces/blob/simulation_interfaces/msg/SimulationState.msg)

This service will be handled by `ROS 2 Simulator manager`.

The transition from *PLAYING* or *PAUSED* to *STOPPED* will trigger level reloading.

The transition from *PLAYING* to *PAUSED* will ask the default physics scene to be disabled. It will stop movement of all PhysX articulations, rigid bodies (both kinematic and simulated), and characters, but some animations will be played. The transition from *PAUSED* to *PLAYING* will do the opposite.

> The transition from *PLAYING*, *PAUSED*, or *STOPPED* to *QUITTING* will call:

Reviewer: I copy my question from the other PR, just to remember about it:
Is there no way to close simulation in a bit more clean way? Or is this approach used elsewhere or is suggested?

Reviewer 2: It was for laughs.

Author: I kept it, so more people laugh.

This service will be handled by `ROS 2 Simulator manager`. If transition is in progress (e.g. reloading level or despawning), the old state will be returned.

# Deprecated components in ROS 2 Gem

During this effort some components will be retired, and some moved outside of scope of supported and canonical repos. This is primarily caused by ROS 2 community retiring `gazebo_msgs`.

### ROS2SpawnerComponent

The `ROS2SpawnerComponent` can co-exist with managers for *Simulation Interfaces* proposed in this RFC, but ROS 2 dependencies (namely `gazebo_msgs`) will be removed from upcoming ROS 2 distribution (`Kilted Kaiju`). The component will be marked deprecated (instead of completely removed from the codebase) to make the transition to Simulation Interfaces an easier experience. The component will be hidden away from users with the *CMake variable*. This variable will be first set to *enabled* by default, and toggled to *disabled* after the deprecation period into the retirement period. Finally, the component will be completely removed from ROS 2 Gem. 

> **Note:** Additional message during the *configure* step will be shown to the user if `gazebo_msgs` package is not detected, but the *variable* is set *true*.

Reviewer: ```suggestion
**Note:** Additional message during the *configure* step will be shown to the user if `gazebo_msgs` package is not detected, but the *variable* is set to *enabled*.
```

Users of `Kilted Kaiju` (and newer) ROS 2 releases will be able to use the deprecated component by building the package manually in their custom ROS 2 workspace. This involves:
- cloning and building the latest [gazebo_ros_pkgs/gazebo_msgs](https://github.com/ros-simulation/gazebo_ros_pkgs/tree/3.9.0/gazebo_msgs)
- sourcing custom workspace
- setting CMake flag
- building ROS2 gem from source.

The deprecation period and the retirement period of the component are yet to be decided.

### ROS2ContactSensorComponent

The `ROS2ContactSensorComponent` is independent of *Simulation Interfaces* proposed in this RFC, but ROS 2 dependencies (namely `gazebo_msgs`) will be removed from upcoming ROS 2 distribution (`Kilted Kaiju`). In particular, this component uses [`ContactState` message](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/3.9.0/gazebo_msgs/msg/ContactState.msg). The component will be hidden away from users with the *CMake variable*. This variable will be first set to *enabled* by default, and toggled to *disabled* after the deprecation period into the retirement period. Finally, the component will be completely removed from ROS 2 Gem. 

> **Note:** Additional message during the *configure* step will be shown to the user if `gazebo_msgs` package is not detected, but the *variable* is set *true*.

Reviewer: For consistency with paragraph above.
```suggestion
**Note:** Additional message during the *configure* step will be shown to the user if `gazebo_msgs` package is not detected, but the *variable* is set to *enabled*.
```

Users of `Kilted Kaiju` (and newer) ROS 2 releases will be able to use the deprecated component by building the package manually in their custom ROS 2 workspace. This involves:
- cloning and building the latest [gazebo_ros_pkgs/gazebo_msgs](https://github.com/ros-simulation/gazebo_ros_pkgs/tree/3.9.0/gazebo_msgs)
- sourcing custom workspace

> - setting CMake flag

Reviewer 2: Flag will be managed by existance of gazebo_msgs.
```suggestion
```

<!-- - Explain any disadvantages for someone to use this feature -->
Significant increase in the code base size for ROS 2 Gem.

### How will this be implemented or integrated into the O3DE environment?
<!-- - Explain how this will be integrated within codebase of O3DE and any special library or technical stack requirements. -->
The *Simulation Interfaces* feature will build over the ROS 2 Gem. In particular, three separate tools for managing the *spawnables*, the simulation and the named poses will be implemented as O3DE *System Components*.  The tools will be callable using O3DE bus system and via ROS 2 middleware. The latter will be decoupled from the *Simulation Interfaces* implementation to ensure the system can be used with different frameworks in the future. More details about the design are given in section [ROS 2 API](#ros-2-api).

### Are there any alternatives to this feature?
<!-- - Provide any other designs that have been considered. Explain what the impact might be to not doing this.
- If there is any prior art or approaches with other frameworks in the same domain, explain how they may have solved this problem or implemented this feature. -->

> The effort needs to be taken to adjust the existing code to the deprecation of Gazebo Classic and `gazebo_msgs`. *ROS 2 Simulation Interfaces* are the upcoming simulation standard that was primarily designed by O3DE *sig-simulation* developers. It will be integrated into all simulation engines and, therefore, it makes no sense to 

Reviewer: Missing end of the sentence here.

> The effort needs to be taken to adjust the existing code to the deprecation of Gazebo Classic and `gazebo_msgs`. *ROS 2 Simulation Interfaces* are the upcoming simulation standard that was primarily designed by O3DE *sig-simulation* developers. It will be integrated into all simulation engines and, therefore, it makes no sense to 

Reviewer 2: ```suggestion
The effort needs to be taken to adjust the existing code to the deprecation of Gazebo Classic and `gazebo_msgs`. *ROS 2 Simulation Interfaces* are the upcoming simulation standard that was primarily designed by O3DE *sig-simulation* developers. It will be integrated into all simulation engines and, therefore, it makes no sense to omit introduction. 
```
